File: clojure/collatz-conjecture/src/collatz_conjecture.clj
(ns collatz-conjecture)

(defn divide-by-2 [x] (/ x 2))

(defn three-x-plus-1 [x] (+ 1 (* x 3)))

(defn collatz [num]
  (if (not (pos? num))
    (throw (Exception. "Collatz calculation requires positive numbers"))
    (loop [count 0 current num]
      (if (= current 1)
        count
        (recur (inc count) (if (even? current) (divide-by-2 current) (three-x-plus-1 current)))
      )
    )
  )
)

File: lua/collatz-conjecture/collatz-conjecture.lua
return function(number)
  if number <= 0 then error'Only positive numbers are allowed' end
  if number == 1 then return 0 end

  local count = 0

  local function is_even(n)
    return n % 2 == 0
  end

  local function collatz(n)
    count = count + 1

    local num = is_even(n) and (n / 2) or ((n * 3) + 1)

    if num == 1 then
      return count
    else
      return collatz(num)
    end
  end

  return collatz(number)
end
